File: old_version/C/第4章 模式匹配/4.5 KMP/kmp.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int* getNexts(char *p) {
    int n = strlen(p);
    int *next = (int *)malloc(sizeof(int) * n);
    memset(next, 0, sizeof(int) * n);
    int j = 0;
    for(int i = 2; i < n; i++) {
        while(j != 0 && p[j] != p[i-1]) {
            j = next[j];
        }
        if(p[j] == p[i-1]) {
            j++;
        }
        next[i] = j;
    }
    return next;
}

int kmp(char *s, char *p) {
    int *next = getNexts(p);
    int j = 0;
    int sLen = strlen(s);
    int pLen = strlen(p);

    for(int i = 0; i < sLen; i++) {
        while(j > 0 && s[i] != p[j]) {
            j = next[j];
        }
        if(s[i] == p[j]) {
            j++;
        }
        if(j == pLen) {
            free(next);
            return i - pLen + 1;
        }
    }

    free(next);
    return -1;
}

int main() {
    char *s = "ATGTGAGCTGGTGTGTGCFAA";
    char *p = "GTGTGCF";
    printf("%d\n", kmp(s, p));
    return 0;
}

File: code/Java/第16章 模式匹配/16.5 KMP/KMP.java
public class KMP {
    public static void main(String[] args) {
        String s = "ATGTGAGCTGGTGTGTGCFAA";
        String p = "GTGTGCF";
        System.out.println(kmp(s, p));
    }

    public static int[] getNexts(String p) {
        int n = p.length();
        int[] next = new int[n];
        int j = 0;
        for(int i = 2; i < n; i++) {
            while(j != 0 && p.charAt(j) != p.charAt(i-1)) {
                j = next[j];
            }
            if(p.charAt(j) == p.charAt(i-1)) {
                j++;
            }
            next[i] = j;
        }
        return next;
    }

    public static int kmp(String s, String p) {
        int[] next = getNexts(p);
        int j = 0;
        for(int i = 0; i < s.length(); i++) {
            while(j > 0 && s.charAt(i) != p.charAt(j)) {
                j = next[j];
            }
            if(s.charAt(i) == p.charAt(j)) {
                j++;
            }
            if(j == p.length()) {
                return i - p.length() + 1;
            }
        }
        return -1;
    }
}
